### ISSUE TITLE
[AMD][kernel][Tier 5] FlyDSL MoE backward dweight grouped-GEMM for training (split from #32)

### ISSUE THREAD

**Issue by sunxxuns** [amdpilot-org/FlyDSL#34]
## Goal

Use FlyDSL to generate the **MoE backward dweight kernel** for
training: grouped backward GEMM (input^T @ output_grad) → per-expert weight gradient. One of 2 sibling issues splitting the former
FlyDSL #32 (MoE backward) — the joint dweight+dactivation umbrella
hit `formulation_stuck` overnight on the autonomous loop.

The weight-update half of MoE backward — needed by every optimizer step.

Reuses cached **`amd/Kimi-K2.5-MXFP4`** (384 experts) +
**`amd/Qwen3.5-397B-A17B-MXFP4`** (128 experts).

## Hardware / workload

```
Hardware:    8 × MI355X (single node, gfx950), EP=8 + TP=1
Models:      amd/Kimi-K2.5-MXFP4 + amd/Qwen3.5-397B-A17B-MXFP4
Shape:       per-expert backward GEMM with grouped layout
             (groups=num_experts_local)
Baseline:    per-expert dense bwd-weight GEMM loop (current Python-loop fallback)
Target:      grouped-bwd kernel ms < 0.3 × per-expert-loop baseline at
             parity_tol < 5e-3 on the gradient tensor.
```

## Scope (single kernel)

This issue ships exactly one kernel.

1. Add `kernels/moe_gemm_2stage_dweight.py` mirroring the existing
   forward layout in `moe_gemm_2stage*.py`.
2. Companion MXFP4-scaled variant
   `moe_gemm_2stage_dweight_mxscale_gfx1250.py` if the FP4 path
   needs the same fusion (mirrors the forward family naming).
3. Parity test in `tests/kernels/test_moe_dweight.py` covering
   experts ∈ {8, 32, 128}, top-k ∈ {2, 8}, intermediate ∈ {2048, 4096}.
4. Benchmark in `tests/kernels/bench_moe_dweight.py` reporting
   `kernel_ms` for the Kimi-K2.5 + Qwen3.5-397B shapes.

## Acceptance

- Parity test passes.
- Benchmark reports `dweight_kernel_ms < 0.3 × loop_baseline`.
- Autograd hookup (the sibling issue + this one together) is
  validated separately in a 2-step backprop sanity test once both
  halves land — out of scope for this issue.

## Why this is a single-kernel issue

The full MoE backward umbrella was in amdpilot-org/FlyDSL#32 — that
landed `formulation_stuck` overnight. This issue is one
kernel + tests + bench. The autograd wiring (`torch.autograd.Function`
that calls both halves) is a tiny follow-up issue once both halves ship.

Sibling: the other of dweight / dactivation.

**Comment by jhinpan** [comment]
**AMDPilot — accumulated progress on this issue**

_Auto-generated from successive trials. Latest update: 2026-05-23T08:56:06Z by `issue-FlyDSL-34-r7`._

### Verified infra fixes (apply during Stage 0)

- Confirmed GPU allocator viability via direct torch.bmm test on device='cuda'
- Validated runtime environment with torch version 2.9.0a0+git7bcbafe


### Working approaches and their best metric

_(none yet — when a trial measurably improves the metric, the change lands here.)_


### Approaches that did NOT help — don't retry these

- torch._grouped_mm failed with 'grouped gemm is not supported on ROCM'
- Importing 'aiter' library failed (exit code 1), forcing bypass to raw PyTorch
- Redundant filesystem probes (find /, ls -la) instead of installing missing 'amdpilot' package
- Assuming repo root is /workspace rather than a subdirectory caused repeated discovery failures
- Overwriting test harness files (probe_dweight_kernel.py) without addressing the 0.125 parity gap on kimi_k2p5
- Shifting focus to bmm_chunked variants and raw bmm speed instead of fixing correctness gaps


### Open ideas to try next

- Pre-scan torch build metadata for rocm flags to auto-flag unsupported ops
- Execute mandatory profiling and save traces to /workspace/traces/ before code changes
- Inspect moe_gemm_2stage_wmma_gfx1250.py for WMMA adaptations to MI355X target
- Inject 'repo_subdir' from issue packet into agent context to prevent root directory assumptions
- Fix harness sanitizer to block pipeline launch when AMDPILOT_METRIC block is missing
- Address node mislabeling: ensure GPU jobs do not schedule on 'cpu-only' nodes like mia1-p02-g46
- Enhance Supervisor validation to auto-re-run benchmarks with parity > 0.0 before allowing new experiments
- Re-verify large model configurations (kimi_k2p5) and tighten test assertions before optimization


_amdpilot bot. Edit between the markers to add/correct fixes; the next update will preserve your changes. To disable updates on this issue, replace the body with the marker line only._

**Comment by sunxxuns** [comment]
## Closing as not tractable by the current autonomous agent (round 2, 2026-05-23)

Chronic-zero-win lineage: 4+ retries in 12 h with 0 completions. Triage: MoE bwd dweight — needs MFMA expertise the agent doesn't have.

The agent (Kimi-K2.6) cannot make iterative progress on this scope alone; every
retry lands on the same blocker. Closing so the fleet stops burning GPU cycles.

**To re-open**: decompose into a single-shape sweep or single-kernel-with-measurable-metric task, or assign to a human kernel engineer.
